homeassistant/components/light/yeelight.py

```diff
@@ -115,16 +127,44 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
                                    discovery_info['properties']['mac'])
         device = {'name': name, 'ipaddr': discovery_info['host']}
 
-        lights.append(YeelightLight(device, DEVICE_SCHEMA({})))
+        light = YeelightLight(device, DEVICE_SCHEMA({}))
+        lights.append(light)
+        hass.data[DATA_KEY][name] = light
     else:
         for ipaddr, device_config in config[CONF_DEVICES].items():
-            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
+            name = device_config[CONF_NAME]
+            _LOGGER.debug("Adding configured %s", name)
 
-            device = {'name': device_config[CONF_NAME], 'ipaddr': ipaddr}
-            lights.append(YeelightLight(device, device_config))
+            device = {'name': name, 'ipaddr': ipaddr}
+            light = YeelightLight(device, device_config)
+            lights.append(light)
+            hass.data[DATA_KEY][name] = light
 
     add_devices(lights, True)
 
+    def service_handler(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_devices = [dev for dev in hass.data[DATA_KEY].values()
+                              if dev.entity_id in entity_ids]
+        else:
+            target_devices = hass.data[DATA_KEY].values()
+
+        for target_device in target_devices:
+            if service.service == SERVICE_SET_MODE:
+                target_device.set_mode(**params)
+
+    service_schema_set_mode = YEELIGHT_SERVICE_SCHEMA.extend({
+        vol.Required(ATTR_MODE):
+            vol.In([mode.name.lower() for mode in PowerMode])
+    })
+    hass.services.register(
```

Comment:
You're registering the new service with the `light` domain. Is that intentional? I think this should be a service of the `yeelight` domain, since only yeelight supports it. What happens if I call this service with the entity ID of some other random light?

Reply:
This was one of my learnings from past reviews. I will check it for safety!

Reply:
I'm pretty sure the `light` domain is correct here. The `light/lifx` component and `media_player/snapcast` does the same. I cannot find any side effects.

Reply:
The platform which responds to these custom commands is encoded in the service name, and that's the way it is done currently. If no one objects this can be merged to my opinion.